spirv_msl.cpp

```diff
@@ -7435,7 +8034,7 @@ string CompilerMSL::to_struct_member(const SPIRType &type, uint32_t member_type_
 	SPIRType row_major_physical_type;
 	const SPIRType *declared_type = &physical_type;
 
-	if (member_is_packed_physical_type(type, index))
+	if (member_is_packed_physical_type(type, index) && validate_member_packing_rules_msl(type, index))
```

Comment:
Why validate member packing rules again? This should be done already by the time we come here.

Reply:
There is a regression here:
```
#version 450
#extension GL_EXT_scalar_block_layout : require

layout(set = 0, binding = 0, scalar) buffer SSBO
{
	vec3 v[16];
};

void main()
{
	v[1] = v[0];
}
```

ends up as:

```
#pragma clang diagnostic ignored "-Wmissing-prototypes"
#pragma clang diagnostic ignored "-Wmissing-braces"

#include <metal_stdlib>
#include <simd/simd.h>

using namespace metal;

template<typename T, size_t Num>
struct spvUnsafeArray
{
    T elements[Num ? Num : 1];
    
    thread T& operator [] (size_t pos) thread
    {
        return elements[pos];
    }
    constexpr const thread T& operator [] (size_t pos) const thread
    {
        return elements[pos];
    }
    
    device T& operator [] (size_t pos) device
    {
        return elements[pos];
    }
    constexpr const device T& operator [] (size_t pos) const device
    {
        return elements[pos];
    }
    
    constexpr const constant T& operator [] (size_t pos) const constant
    {
        return elements[pos];
    }
    
    threadgroup T& operator [] (size_t pos) threadgroup
    {
        return elements[pos];
    }
    constexpr const threadgroup T& operator [] (size_t pos) const threadgroup
    {
        return elements[pos];
    }
};

struct SSBO
{
    packed_spvUnsafeArray<float3, 16> v;
};

kernel void main0(device SSBO& _13 [[buffer(0)]])
{
    spvUnsafeArray<float3, 16>(_13.v[1]) = float3(spvUnsafeArray<float3, 16>(_13.v[0]));
}
```
it should be using `packed_float3 v[16]` instead. There's a lot of packing tests in `shaders-msl-no-opt/packing/`.

Reply:
You're right. I added two tests to that folder:
- `array-of-vec3.comp` which is basically your example
- `array-of-vec4.comp` to verify `spvUnsafeArray` is still used when packing is not necessary.